sdks/python/apache_beam/io/gcp/bigtableio.py

```diff
@@ -36,43 +36,170 @@
 
+__all__ = ['BigtableSource', 'BigtableWrite']
 
-class _BigTableWriteFn(beam.DoFn):
-  """ Creates the connector can call and add_row to the batcher using each
-  row in beam pipe line
-  Args:
-    project_id(str): GCP Project ID
-    instance_id(str): GCP Instance ID
-    table_id(str): GCP Table ID
 
-  """
+class BigtableSource(BoundedSource):
+  def __init__(self, project_id, instance_id, table_id, filter_=None):
+    """ Constructor of the Read connector of Bigtable
 
-  def __init__(self, project_id, instance_id, table_id):
-    """ Constructor of the Write connector of Bigtable
     Args:
-      project_id(str): GCP Project of to write the Rows
-      instance_id(str): GCP Instance to write the Rows
-      table_id(str): GCP Table to write the `DirectRows`
+      project_id: [string] GCP Project of to write the Rows
+      instance_id: [string] GCP Instance to write the Rows
+      table_id: [string] GCP Table to write the `DirectRows`
+      filter_: [RowFilter] Filter to apply to cells in a row.
     """
-    super(_BigTableWriteFn, self).__init__()
+    super(self.__class__, self).__init__()
     self.beam_options = {'project_id': project_id,
                          'instance_id': instance_id,
-                         'table_id': table_id}
+                         'table_id': table_id,
+                         'filter_': filter_}
+    self.sample_row_keys = None
+    self.table = None
+    self.row_count = Metrics.counter(self.__class__.__name__, 'Row count')
+
+  def __getstate__(self):
+    return self.beam_options
+
+  def __setstate__(self, options):
+    self.beam_options = options
+    self.table = None
+    self.sample_row_keys = None
+    self.row_count = Metrics.counter(self.__class__.__name__, 'Row count')
+
+  def _get_table(self):
+    if self.table is None:
+      self.table = Client(project=self.beam_options['project_id'])\
+                      .instance(self.beam_options['instance_id'])\
+                      .table(self.beam_options['table_id'])
+    return self.table
+
+  def get_sample_row_keys(self):
+    """ Get a sample of row keys in the table.
+
+    The returned row keys will delimit contiguous sections of the table of
+    approximately equal size, which can be used to break up the data for
+    distributed tasks like mapreduces.
+    :returns: A cancel-able iterator. Can be consumed by calling ``next()``
+    			  or by casting to a :class:`list` and can be cancelled by
+    			  calling ``cancel()``.
+    """
+    if self.sample_row_keys is None:
+      self.sample_row_keys = list(self._get_table().sample_row_keys())
+    return self.sample_row_keys
+
+  def get_range_tracker(self, start_position=b'', stop_position=b''):
+    if stop_position == b'':
+      return LexicographicKeyRangeTracker(start_position)
+    else:
+      return LexicographicKeyRangeTracker(start_position, stop_position)
+
+  def estimate_size(self):
+    return list(self.get_sample_row_keys())[-1].offset_bytes
```

Comment:
`get_sample_row_keys()` is already a list.  No need to create another `list()` here.

Reply:
Done.